File: app/src/main/java/com/mainstreetcode/teammate/fragments/main/TournamentDetailFragment.java
package com.mainstreetcode.teammate.fragments.main;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.design.widget.TabLayout;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AlertDialog;
import android.transition.AutoTransition;
import android.transition.TransitionManager;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.mainstreetcode.teammate.R;
import com.mainstreetcode.teammate.adapters.TournamentRoundAdapter;
import com.mainstreetcode.teammate.adapters.viewholders.EmptyViewHolder;
import com.mainstreetcode.teammate.adapters.viewholders.ModelCardViewHolder;
import com.mainstreetcode.teammate.adapters.viewholders.TeamViewHolder;
import com.mainstreetcode.teammate.adapters.viewholders.UserViewHolder;
import com.mainstreetcode.teammate.baseclasses.MainActivityFragment;
import com.mainstreetcode.teammate.model.Competitive;
import com.mainstreetcode.teammate.model.Competitor;
import com.mainstreetcode.teammate.model.Team;
import com.mainstreetcode.teammate.model.Tournament;
import com.mainstreetcode.teammate.model.User;
import com.mainstreetcode.teammate.util.ModelUtils;
import com.mainstreetcode.teammate.viewmodel.gofers.Gofer;

import static android.support.design.widget.TabLayout.MODE_FIXED;
import static android.support.design.widget.TabLayout.MODE_SCROLLABLE;

public class TournamentDetailFragment extends MainActivityFragment {

    public static final String ARG_TOURNAMENT = "role";

    private Tournament tournament;
    private ViewPager viewPager;
    private TabLayout tabLayout;
    private EmptyViewHolder viewHolder;

    public static TournamentDetailFragment newInstance(Tournament tournament) {
        TournamentDetailFragment fragment = new TournamentDetailFragment();
        Bundle args = new Bundle();

        args.putParcelable(ARG_TOURNAMENT, tournament);
        fragment.setArguments(args);
        fragment.setEnterExitTransitions();

        return fragment;
    }

    @Override
    @SuppressWarnings("ConstantConditions")
    public String getStableTag() {
        return Gofer.tag(super.getStableTag(), getArguments().getParcelable(ARG_TOURNAMENT));
    }

    @Override
    @SuppressWarnings("ConstantConditions")
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        tournament = getArguments().getParcelable(ARG_TOURNAMENT);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_games_parent, container, false);
        viewPager = root.findViewById(R.id.view_pager);
        tabLayout = root.findViewById(R.id.tab_layout);
        viewHolder = new EmptyViewHolder(root, R.drawable.ic_score_white_24dp, R.string.tournament_games_desc);

        viewPager.setAdapter(new TournamentRoundAdapter(tournament, getChildFragmentManager()));
        viewPager.setCurrentItem(tournament.getCurrentRound());

        setUpWinner((ViewGroup) root, tournament.getNumRounds());

        return root;
    }

    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        super.onPrepareOptionsMenu(menu);
        boolean hasPrivilegedRole = localRoleViewModel.hasPrivilegedRole();
        menu.findItem(R.id.action_edit).setVisible(hasPrivilegedRole);
        menu.findItem(R.id.action_delete).setVisible(hasPrivilegedRole);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.fragment_tournament_detail, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_edit:
                showFragment(TournamentEditFragment.newInstance(tournament));
                break;
            case R.id.action_standings:
                showFragment(StatDetailFragment.newInstance(tournament));
                break;
            case R.id.action_delete:
                Context context = getContext();
                if (context == null) return true;
                new AlertDialog.Builder(context).setTitle(getString(R.string.delete_tournament_prompt))
                        .setPositiveButton(R.string.yes, (dialog, which) -> deleteTournament())
                        .setNegativeButton(R.string.no, (dialog, which) -> dialog.dismiss())
                        .show();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onResume() {
        super.onResume();
        int rounds = tournament.getNumRounds();
        User user = userViewModel.getCurrentUser();
        Team team = tournament.getHost();
        disposables.add(localRoleViewModel.getRoleInTeam(user, team).subscribe(this::togglePersistentUi, emptyErrorHandler));
        disposables.add(tournamentViewModel.checkForWinner(tournament).subscribe(changed -> setUpWinner((ViewGroup) getView(), rounds), defaultErrorHandler));
    }

    @Override
    public void togglePersistentUi() {
        updateFabIcon();
        setToolbarTitle(getString(R.string.tournament_fixtures));
        requireActivity().invalidateOptionsMenu();
        super.togglePersistentUi();
    }

    @Override
    public void onDestroyView() {
        viewPager = null;
        tabLayout = null;
        viewHolder = null;
        super.onDestroyView();
    }

    @Override
    @StringRes
    protected int getFabStringResource() { return R.string.add_tournament_competitors; }

    @Override
    @DrawableRes
    protected int getFabIconResource() { return R.drawable.ic_group_add_white_24dp; }

    @Override
    public boolean showsFab() {
        return localRoleViewModel.hasPrivilegedRole() && !tournament.hasCompetitors();
    }

    @Override
    public void onClick(View v) {
        if (v.getId() == R.id.fab) showFragment(CompetitorsFragment.newInstance(tournament));
    }

    private void deleteTournament() {
        disposables.add(tournamentViewModel.delete(tournament).subscribe(this::onTournamentDeleted, defaultErrorHandler));
    }

    private void onTournamentDeleted(Tournament deleted) {
        showSnackbar(getString(R.string.deleted_team, deleted.getName()));
        removeEnterExitTransitions();

        Activity activity;
        if ((activity = getActivity()) == null) return;

        activity.onBackPressed();
    }

    @SuppressWarnings("unchecked")
    private void setUpWinner(@Nullable ViewGroup root, int prevAdapterCount) {
        if (root == null) return;

        TextView winnerText = root.findViewById(R.id.winner);
        ViewGroup winnerView = root.findViewById(R.id.item_container);
        PagerAdapter adapter = root.<ViewPager>findViewById(R.id.view_pager).getAdapter();

        if (prevAdapterCount != tournament.getNumRounds() && adapter != null)
            adapter.notifyDataSetChanged();

        TransitionManager.beginDelayedTransition(root, new AutoTransition()
                .addTarget(tabLayout)
                .addTarget(viewPager)
                .addTarget(winnerView)
                .addTarget(winnerText));

        boolean hasCompetitors = tournament.getNumCompetitors() > 0;
        tabLayout.setTabMode(tournament.getNumRounds() > 4 ? MODE_SCROLLABLE : MODE_FIXED);
        tabLayout.setVisibility(hasCompetitors ? View.VISIBLE : View.GONE);
        tabLayout.setupWithViewPager(viewPager);
        viewHolder.setColor(R.color.dark_grey);
        viewHolder.toggle(!hasCompetitors);

        Competitor winner = tournament.getWinner();
        if (winner.isEmpty()) return;

        Competitive competitive = winner.getEntity();

        ModelCardViewHolder viewHolder = competitive instanceof User
                ? new UserViewHolder(winnerView, user -> {})
                : competitive instanceof Team
                ? new TeamViewHolder(winnerView, team -> {})
                : null;

        if (viewHolder == null) return;
        viewHolder.bind(competitive);

        winnerText.setVisibility(View.VISIBLE);
        winnerView.setVisibility(View.VISIBLE);

        winnerText.setText(ModelUtils.processString(getString(R.string.tournament_winner)));
    }
}

File: app/src/main/java/com/mainstreetcode/teammate/fragments/main/TournamentDetailFragment.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Adetunji Dahunsi
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.mainstreetcode.teammate.fragments.main

import android.os.Bundle
import android.transition.AutoTransition
import android.transition.TransitionManager
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.core.os.bundleOf
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager.widget.ViewPager
import com.google.android.material.tabs.TabLayout.MODE_FIXED
import com.google.android.material.tabs.TabLayout.MODE_SCROLLABLE
import com.mainstreetcode.teammate.R
import com.mainstreetcode.teammate.adapters.TeamAdapter
import com.mainstreetcode.teammate.adapters.TournamentRoundAdapter
import com.mainstreetcode.teammate.adapters.UserAdapter
import com.mainstreetcode.teammate.adapters.viewholders.EmptyViewHolder
import com.mainstreetcode.teammate.adapters.viewholders.TeamViewHolder
import com.mainstreetcode.teammate.adapters.viewholders.UserViewHolder
import com.mainstreetcode.teammate.baseclasses.TeammatesBaseFragment
import com.mainstreetcode.teammate.databinding.FragmentGamesParentBinding
import com.mainstreetcode.teammate.model.Competitor
import com.mainstreetcode.teammate.model.Team
import com.mainstreetcode.teammate.model.Tournament
import com.mainstreetcode.teammate.model.User
import com.mainstreetcode.teammate.util.ErrorHandler
import com.mainstreetcode.teammate.util.processEmoji
import com.mainstreetcode.teammate.viewmodel.gofers.Gofer

class TournamentDetailFragment : TeammatesBaseFragment(R.layout.fragment_games_parent) {

    private lateinit var tournament: Tournament
    private lateinit var competitor: Competitor

    private var binding: FragmentGamesParentBinding? = null
    private var viewHolder: EmptyViewHolder? = null

    internal var gamesRecycledViewPool: RecyclerView.RecycledViewPool = RecyclerView.RecycledViewPool()
        private set

    override val showsFab: Boolean get() = localRoleViewModel.hasPrivilegedRole() && !tournament.hasCompetitors()

    internal fun pending(competitor: Competitor): TournamentDetailFragment = apply {
        arguments!!.putParcelable(ARG_COMPETITOR, competitor)
    }

    override val stableTag: String
        get() =
            Gofer.tag(super.stableTag, arguments!!.getParcelable(ARG_TOURNAMENT)!!)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val args = arguments
        tournament = args!!.getParcelable(ARG_TOURNAMENT)!!
        competitor = args.getParcelable(ARG_COMPETITOR) ?: Competitor.empty()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) = FragmentGamesParentBinding.bind(view).run {
        super.onViewCreated(view, savedInstanceState)
        defaultUi(
                toolbarTitle = getString(R.string.tournament_fixtures),
                toolBarMenu = R.menu.fragment_tournament_detail,
                fabIcon = R.drawable.ic_group_add_white_24dp,
                fabText = R.string.add_tournament_competitors,
                fabShows = showsFab
        )
        viewHolder = EmptyViewHolder(view, R.drawable.ic_score_white_24dp, R.string.tournament_games_desc)

        viewPager?.adapter = TournamentRoundAdapter(tournament, childFragmentManager)
        viewPager?.currentItem = tournament.currentRound

        binding = this

        setUpWinner(tournament.numRounds)
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        super.onPrepareOptionsMenu(menu)
        val hasPrivilegedRole = localRoleViewModel.hasPrivilegedRole()
        menu.findItem(R.id.action_edit)?.isVisible = hasPrivilegedRole
        menu.findItem(R.id.action_delete)?.isVisible = hasPrivilegedRole
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean = when (item.itemId) {
        R.id.action_edit -> navigator.push(TournamentEditFragment.newInstance(tournament))
        R.id.action_standings -> navigator.push(StatDetailFragment.newInstance(tournament))
        R.id.action_delete -> AlertDialog.Builder(requireContext()).setTitle(getString(R.string.delete_tournament_prompt))
                .setPositiveButton(R.string.yes) { _, _ -> deleteTournament() }
                .setNegativeButton(R.string.no) { dialog, _ -> dialog.dismiss() }
                .show().let { true }
        else -> super.onOptionsItemSelected(item)
    }

    override fun onResume() {
        super.onResume()
        checkCompetitor()
        watchForRoleChanges(tournament.host) { updateUi(fabShows = showsFab, toolbarInvalidated = true) }

        val rounds = tournament.numRounds
        disposables.add(tournamentViewModel.checkForWinner(tournament)
                .subscribe({ setUpWinner(rounds) }, defaultErrorHandler::invoke))
    }

    override fun onDestroyView() {
        binding = null
        viewHolder = null
        super.onDestroyView()
    }

    override fun onClick(view: View) {
        if (view.id == R.id.fab) navigator.push(CompetitorsFragment.newInstance(tournament))
    }

    private fun checkCompetitor() {
        if (competitor.isEmpty || competitor.isAccepted) return
        if (restoredFromBackStack)
        // Don't prompt for the same competitor multiple times.
            disposables.add(competitorViewModel.updateCompetitor(competitor).subscribe(this::promptForCompetitor, ErrorHandler.EMPTY::invoke))
        else
            promptForCompetitor()
    }

    private fun promptForCompetitor() {
        if (competitor.isEmpty || competitor.isAccepted) return
        transientBarDriver.showChoices { choiceBar ->
            choiceBar.setText(getString(R.string.tournament_accept))
                    .setPositiveText(getText(R.string.accept))
                    .setNegativeText(getText(R.string.decline))
                    .setPositiveClickListener(View.OnClickListener { respond(true) })
                    .setNegativeClickListener(View.OnClickListener { respond(false) })
        }
    }

    private fun respond(accept: Boolean) {
        transientBarDriver.toggleProgress(true)
        disposables.add(competitorViewModel.respond(competitor, accept)
                .subscribe({ transientBarDriver.toggleProgress(false) }, defaultErrorHandler::invoke))
    }

    private fun deleteTournament() {
        disposables.add(tournamentViewModel.delete(tournament).subscribe(this::onTournamentDeleted, defaultErrorHandler::invoke))
    }

    private fun onTournamentDeleted(deleted: Tournament) {
        transientBarDriver.showSnackBar(getString(R.string.deleted_team, deleted.name))
        removeEnterExitTransitions()
        requireActivity().onBackPressed()
    }

    private fun setUpWinner(prevAdapterCount: Int) = binding?.run {
        val winnerText = root.findViewById<TextView>(R.id.winner)
        val winnerView = root.findViewById<ViewGroup>(R.id.item_container)
        val adapter = root.findViewById<ViewPager>(R.id.view_pager).adapter

        if (prevAdapterCount != tournament.numRounds && adapter != null)
            adapter.notifyDataSetChanged()

        TransitionManager.beginDelayedTransition(innerCoordinator, AutoTransition()
                .addTarget(tabLayout)
                .addTarget(viewPager)
                .addTarget(winnerView)
                .addTarget(winnerText))

        val hasCompetitors = tournament.numCompetitors > 0
        tabLayout.tabMode = if (tournament.numRounds > 4) MODE_SCROLLABLE else MODE_FIXED
        tabLayout.visibility = if (hasCompetitors) View.VISIBLE else View.GONE
        tabLayout.setupWithViewPager(viewPager)
        viewHolder?.setColor(R.attr.alt_empty_view_holder_tint)
        viewHolder?.toggle(!hasCompetitors)

        val winner = tournament.winner
        if (winner.isEmpty) return

        when (val competitive = winner.entity) {
            is User -> UserViewHolder(winnerView, UserAdapter.AdapterListener.asSAM { showCompetitor(winner) }).apply { bind(competitive) }
            is Team -> TeamViewHolder(winnerView, TeamAdapter.AdapterListener.asSAM { showCompetitor(winner) }).apply { bind(competitive) }
            else -> return
        }

        winnerText.visibility = View.VISIBLE
        winnerView.visibility = View.VISIBLE

        winnerText.text = getString(R.string.tournament_winner).processEmoji()
    } ?: Unit

    companion object {

        private const val ARG_TOURNAMENT = "tournament"
        private const val ARG_COMPETITOR = "competitor"

        fun newInstance(tournament: Tournament): TournamentDetailFragment = TournamentDetailFragment().apply {
            arguments = bundleOf(ARG_TOURNAMENT to tournament)
            setEnterExitTransitions()
        }
    }
}
